Lib/asyncio/locks.py

```diff
@@ -212,11 +216,15 @@ def release(self):
             raise RuntimeError('Lock is not acquired.')
 
     def _wake_up_first(self):
-        """Wake up the first waiter who isn't cancelled."""
-        for fut in self._waiters:
-            if not fut.done():
-                fut.set_result(True)
-                break
+        """Wake up the first waiter if it isn't done."""
```

Comment:
What if it is done?  Can there be a situation where you have [fut<done>, fut<not done>] in the _waiters list?

Reply:
There can be indeed, when 2 or more locks are waiting on acquire.
If the first future is done, it means that later on it will either get the lock, or it was cancelled in which case it will wake up the next waiter upon loop's next cycle.
The patch fixes the bug by avoiding multiple done waiters at once and cleaning up the waiters list asap.
The only way multiple done waiters can happen is if 2 futures get cancelled in which case only 1 waiter will wake up, and the second cancelled future will not wake up another one as lock would have already been taken (`if not self._locked: self._wake_up_first()`) or it wasn't taken yet it which case it's still sitting .done() in the waiter's list.

In summary:
.done necessarily means either the lock is immediately taken next loop or if it isn't, (.done occurred because of cancellation) next one wakes up. So if we have a done future not including the one currently running we shouldn't wake up another one as it will cause a double wakeup, hence, the almighty bug.

Reply:
OK, makes sense.  Could you please add this all as a comment to `_wake_up_first`? :)

Reply:
Phew, now that's a long comment, lemme see how can I shorten it while still managing to fully explain it :P

Reply:
Sure, if you have time to write a short comment :)

Sorry for being anal about this. Every once in a while we find a new bug in `asyncio.Lock` or `asyncio.Queue` and it's always a challenge to wrap your head around how they are actually working.  Every line of code is significant, sometimes it's not safe to do a simple refactoring because things break in very non-obvious ways.

Reply:
Nah, all good mate. It took some time to wrap my head around it as well.
Luckily I didn't actually encounter the bug at production but rather found it by reading the implementation which took a couple of hours ;-)

How's this comment? Attempting to be short and to the point as much as you can with synchronization.